### PR TITLE
Introduced service time metrics to OpenSearch-Py client.

### LINKED ISSUE #678: [FEATURE]Add "Service Time" Measurement in opensearch-py
### Is your feature request related to a problem?
Opensearch-py currently lacks service time measurement functionality, hindering performance analysis.

### What solution would you like?
I propose adding service time measurement functionality to opensearch-py, returning service time data conditionally upon user request.

### What alternatives have you considered?

- I haven't found a Pythonic solution for this yet. Any guidance in this regard would be greatly appreciated.
- I'm exploring various approaches and will update with draft PRs soon.

### PR DESCRIPTION
**Pull request by saimedhi**
### Description
Introduced service time metrics to OpenSearch-Py client.

### Issues Resolved
#678 

##### Guide and Tests need to be added for this change.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by saimedhi** [comment]
I'm unsure about my changes to async connections. I've tested them, but I'm uncertain if I've covered all cases. Let's start by applying these changes for requests and urllib3 connections first. For async connections, I'll need to review whether a similar approach to sync connections is suitable or if we should utilize something like aiohttptraceconfig.

File: opensearchpy/connection/http_requests.py [hunk 1]
```diff
@@ -187,9 +189,17 @@ def perform_request(  # type: ignore
             "allow_redirects": allow_redirects,
         }
         send_kwargs.update(settings)
+
+        calculate_service_time = False
```
Comment (dblock) [inline opensearchpy/connection/http_requests.py:193]: This seems very specific. I propose to add `metrics` parameter that would default to `None` or an instance of `Metrics`. I'd like to be able to write 

```py
client = Client(metrics = Metrics())
```

Then the `if` is not necessary.

File: opensearchpy/connection/http_requests.py [hunk 2]
```diff
@@ -187,9 +189,17 @@ def perform_request(  # type: ignore
...
+        if "calculate_service_time" in self.kwargs:
+            calculate_service_time = self.kwargs["calculate_service_time"]
+
+        time_metrics = TimeMetrics()
```
Comment (dblock) [inline opensearchpy/connection/http_requests.py:197]: Create an instance variable for the client called `self.metrics`

File: opensearchpy/connection/http_requests.py [hunk 3]
```diff
@@ -187,9 +189,17 @@ def perform_request(  # type: ignore
...
+
         try:
+            time_metrics.events.server_request_start()
```
Comment (dblock) [inline opensearchpy/connection/http_requests.py:200]: You shouldn't b reaching inside `.events`, it should be something like `self.metrics.request_start()`.

File: opensearchpy/metrics.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+# Modifications Copyright OpenSearch Contributors. See
+# GitHub history for details.
+
+
+import time
+
+from events import Events
+
+
+class TimeMetrics:
```
Comment (dblock) [inline opensearchpy/metrics.py:16]: Create an interface and a class that implements it.

https://www.geeksforgeeks.org/abstract-classes-in-python/

```py
class Metrics(ABC)
   @abstractmethod
   def request_start()
```

```py
class MetricsEvents(Metrics):
   def request_start(self):
      self.start_time ...
```

**Review by dblock — CHANGES_REQUESTED** [review]
Good start. The `if calculate_service_time:` is the anti-pattern we need to fix here and make it more object-oriented. Define a generic `Metrics` interface with methods such as `request_start`, and a specific implementation for it using `Events`, e.g. `EventsMetrics`. See below.

File: opensearchpy/connection/http_requests.py [hunk 5]
```diff
@@ -187,9 +190,14 @@ def perform_request(  # type: ignore
             "allow_redirects": allow_redirects,
         }
         send_kwargs.update(settings)
+        self.is_metrics_none = False
```
Comment (dblock) [inline opensearchpy/connection/http_requests.py:193]: This looks like an anti-pattern. Either `self.metrics` is set, or it's not. Default to not.

Comment (saimedhi) [reply opensearchpy/connection/http_requests.py:193]: @dblock, If it's not used, then [self.metrics.request_start()](https://github.com/opensearch-project/opensearch-py/pull/689/files#diff-1e38589c63abdaf216eafb88d5419b81729acb2c5ccc71d5b4af7a52da105e47R277) and [self.metrics.request_end()](https://github.com/opensearch-project/opensearch-py/pull/689/files#diff-1e38589c63abdaf216eafb88d5419b81729acb2c5ccc71d5b4af7a52da105e47R281) can't be calculated. Are you suggesting we keep [start = time.time()](https://github.com/opensearch-project/opensearch-py/pull/689/files#diff-1e38589c63abdaf216eafb88d5419b81729acb2c5ccc71d5b4af7a52da105e47L245) and [duration = time.time() - start](https://github.com/opensearch-project/opensearch-py/pull/689/files#diff-1e38589c63abdaf216eafb88d5419b81729acb2c5ccc71d5b4af7a52da105e47L274) untouched, and only calculate and return service_time when self.metrics is not None?

Comment (saimedhi) [reply opensearchpy/connection/http_requests.py]: Made the changes as requested. Working on adding tests.

File: opensearchpy/connection/http_requests.py [hunk 6]
```diff
@@ -188,7 +192,11 @@ def perform_request(  # type: ignore
         }
         send_kwargs.update(settings)
         try:
+            if self.metrics is not None:
+                self.metrics.request_start()
             response = self.session.send(prepared_request, **send_kwargs)
+            if self.metrics is not None:
```
Comment (dblock) [inline opensearchpy/connection/http_requests.py:198]: If there's an exception this is never called. Use `finally` to always `request_end`. Add a test.

File: opensearchpy/connection/http_requests.py [hunk 7]
```diff
@@ -244,7 +252,15 @@ def perform_request(  # type: ignore
...
-        return response.status_code, response.headers, raw_data
+        if self.metrics is None:
+            return response.status_code, response.headers, raw_data
+        else:
+            return (
+                response.status_code,
+                response.headers,
+                raw_data,
+                self.metrics.service_time,
+            )
```
Comment (dblock) [inline opensearchpy/connection/http_requests.py:263]: I don't think you should be returning any of the metrics, the caller can look into `.metrics.service_time` if they need to.

File: opensearchpy/metrics.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    def start_time(self) -> float:
+        pass
+
+    @property
+    @abstractmethod
+    def service_time(self) -> float:
+        pass
+
+
+class MetricsEvents(Metrics):
```
Comment (dblock) [inline opensearchpy/metrics.py:36]: Make a new folder/namespace called `metrics`, and separate these classes into individual files.

File: opensearchpy/metrics.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    def start_time(self) -> float:
+        return self._start_time
+
+    @property
+    def service_time(self) -> float:
+        return self._service_time
+
+    def __init__(self) -> None:
+        self.events = Events()
+        self._start_time = 0.0
```
Comment (dblock) [inline opensearchpy/metrics.py:47]: Any reason to not expose these? I would just do `self.start_time`.

These should begin with `None`.

File: opensearchpy/metrics.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+        # Subscribe to the request_start and request_end events
+        self.events.request_start += self._on_request_start
+        self.events.request_end += self._on_request_end
+
+    def request_start(self) -> None:
+        self.events.request_start()
+
+    def _on_request_start(self) -> None:
+        self._start_time = time.perf_counter()
```
Comment (dblock) [inline opensearchpy/metrics.py:58]: Re-assign `self._end_time = None` here. If metrics is reused, you don't want to get wrong time. Add a test.

File: opensearchpy/transport.py [hunk 11]
```diff
@@ -457,6 +478,10 @@ def perform_request(
                     data = self.deserializer.loads(
                         data, headers_response.get("content-type")
                     )
+
+                if self.metrics:
```
Comment (dblock) [inline opensearchpy/transport.py:482]: Why is this needed? You can always look into metrics, can't you?

**Review by dblock — COMMENTED** [review]
Better. Needs tests and some more changes.
